samples/ts-valueentity-shopping-cart/src/shoppingcart.ts

```diff
@@ -0,0 +1,153 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {ValueEntity} from "@lightbend/akkaserverless-javascript-sdk";
+import * as proto from "../lib/generated/proto";
+
+type Context            = ValueEntity.ValueEntityCommandContext;
+
+type State              = proto.com.example.shoppingcart.domain.Cart;
+
+type AddLineItem        = proto.com.example.shoppingcart.AddLineItem
+type RemoveLineItem     = proto.com.example.shoppingcart.RemoveLineItem
+type GetShoppingCart    = proto.com.example.shoppingcart.GetShoppingCart
+type RemoveShoppingCart = proto.com.example.shoppingcart.RemoveShoppingCart
+
+/**
+ * Type definitions.
+ * These types have been generated based on your proto source.
+ * A TypeScript aware editor such as VS Code will be able to leverage them to provide hinting and validation.
+ *
+ * State; the serialisable and persistable state of the entity
+ * @typedef { import("../lib/generated/shoppingcartservice").State } State
+ *
+ * ShoppingCartService; a strongly typed extension of ValueEntity derived from your proto source
+ * @typedef { import("../lib/generated/shoppingcartservice").ShoppingCartService } ShoppingCartService
+ */
+
+/**
+ * @type ShoppingCartService
+ */
+const entity: ValueEntity = new ValueEntity(
+  [
+    "shoppingcart_domain.proto",
+    "shoppingcart_api.proto"
+  ],
+  "com.example.shoppingcart.ShoppingCartService",
+  "shopping-cart",
+  {
+    includeDirs: ["./proto"],
+    serializeFallbackToJson: true
+  }
+);
+
+/*
+ * Here we load the Protobuf types. When setting state, we need to return
+ * protobuf message objects, not just ordinary JavaScript objects, so that the framework can
+ * know how to serialize these objects when they are persisted.
+ *
+ * Note this shows loading them dynamically, they could also be compiled and statically loaded.
+ */
+const pkg = "com.example.shoppingcart.domain.";
+const Cart = entity.lookupType(pkg + "Cart");
+
+/*
+ * Set a callback to create the initial state. This is what is created if there is no
+ * snapshot to load.
+ *
+ * We can ignore the cartId parameter if we want, it's the id of the entity, which is
+ * automatically associated with all events and state for this entity.
+ */
+entity.setInitial(cartId => Cart.create({ items: [] }));
+
+/*
+ * Command handlers. The name of the command corresponds to the name of the rpc call in
+ * the gRPC service that this entity offers.
+ */
+entity.setCommandHandlers({
+  AddItem: addItem,
+  RemoveItem: removeItem,
+  GetCart: getCart,
+  RemoveCart: removeCart
+});
+
+
+/**
+ * Handler for add item commands.
+ */
+function addItem(addItem: AddLineItem, cart: State, ctx: Context) {
+  // Validation:
+  // Make sure that it is not possible to add negative quantities
+  if (addItem.quantity < 1) {
+    ctx.fail("Cannot add negative quantity to item " + addItem.productId);
```

Comment:
We're moving away from the context-based API, to using what's currently the `Reply`-based API here (and becoming `Effect` in the Java SDK, which includes updates to the state, and will likely be mirrored here as well). Maybe this sample should use `replies` now as the preferred API, and also add typing for the return value of handler functions.

Reply:
Updated everywhere.
Can you please check if I am using the `Reply`-based API correctly?
I've never used it before and I was implementing it based on the TCKs examples.
`return {};` (empty response) got replaced by `return replies.noReply();` is that correct?

Reply:
`noReply` should be for no reply at all — different from an empty or default message. It might be translated automatically in this case, but an explicit `replies.message` with an empty is probably correct instead.

Would be good to make `Reply` generic with the message type. The Java SDK is moving on to using `Effect` rather than `Reply`, which I expect may be mirrored here as well. So could be looked at then.

Reply:
> Would be good to make Reply generic with the message type

What do you mean by this? Could you give a simple example?
Does it worth the effort to change the `Reply` in the SDK if we will migrate to `Effect` soon anyways?

Reply:
In the Java SDK, replies are generic, `Reply<T>`, where it's the response message type. So a gRPC method that returns an `Empty` message would have a handler that returns `Reply<Empty>`.

Could be worthwhile having the type for `Reply` now, which will just be transferred over to `Effect` if we do that rename too. `Effect` in the Java SDK is an inner class specific to the type of component — actions, value entities, event sourced entities, and so on, have their own different effects — so something similar would be needed.